proxylib/memcached/parser.go

```diff
@@ -0,0 +1,486 @@
+				return false
+			}
+		}
+		return true
+	}
+
+	if rule.keyPrefix != "" {
+		for _, key := range packetMeta.Keys {
+			if !bytes.HasPrefix(key, rule.keyPrefixBytes) {
+				return false
+			}
+		}
+		return true
+	}
+
+	if rule.keyRegex != "" {
+		for _, key := range packetMeta.Keys {
+			if !rule.regex.Match(key) {
+				return false
+			}
+		}
+		return true
+	}
+
+	log.Debugf("No key rule specified, matching by command")
+	return true
+}
+
+func (rule *Rule) matchCommand(cmd string) bool {
+	_, ok := rule.commands.text[cmd]
+	return ok
+}
+
+func (rule *Rule) matchOpcode(code byte) bool {
+	_, ok := rule.commands.binary[code]
+	return ok
+}
+
+// L7RuleParser parses protobuf L7 rules to and array of Rule
+// May panic
+func L7RuleParser(rule *cilium.PortNetworkPolicyRule) []proxylib.L7NetworkPolicyRule {
+	l7Rules := rule.GetL7Rules()
+	if l7Rules == nil {
+		proxylib.ParseError("Can't get L7 rules", rule)
+	}
+	var rules []proxylib.L7NetworkPolicyRule
+	for _, l7Rule := range l7Rules.GetL7Rules() {
+		var br Rule
+		for k, v := range l7Rule.Rule {
+			switch k {
+			case "command":
+				br.command = v
+				br.commands = MemcacheOpCodeMap[v]
+			case "keyExact":
+				br.keyExact = v
+				br.keyExactBytes = []byte(v)
+			case "keyPrefix":
+				br.keyPrefix = v
+				br.keyPrefixBytes = []byte(v)
+			case "keyRegex":
+				br.keyRegex = v
+				br.regex = regexp.MustCompile(v)
+			default:
+				proxylib.ParseError(fmt.Sprintf("Unsupported key: %s", k), rule)
+			}
+		}
+		if br.command == "" {
+			proxylib.ParseError("command not specified", rule)
+		}
+		log.Debugf("Parsed Rule pair: %v", br)
+		rules = append(rules, &br)
+	}
+	return rules
+}
+
+// ParserFactory implements proxylib.ParserFactory
+type ParserFactory struct{}
+
+// Create creates memcached parser
+func (p *ParserFactory) Create(connection *proxylib.Connection) proxylib.Parser {
+	log.Debugf("ParserFactory: Create: %v", connection)
+	return &Parser{
+		connection:   connection,
+		textParser:   text.ParserFactoryInstance.Create(connection),
+		binaryParser: binary.ParserFactoryInstance.Create(connection),
+	}
+}
+
+// compile time check for interface implementation
+var _ proxylib.ParserFactory = &ParserFactory{}
+
+var memcacheParserFactory *ParserFactory
+
+const (
+	parserName = "memcache"
+)
+
+func init() {
+	log.Info("init(): Registering memcacheParserFactory")
+	proxylib.RegisterParserFactory(parserName, memcacheParserFactory)
+	proxylib.RegisterL7RuleParser(parserName, L7RuleParser)
+}
+
+// Parser implements proxylib.Parser
+type Parser struct {
+	connection   *proxylib.Connection
+	textParser   proxylib.Parser
+	binaryParser proxylib.Parser
+	isBinary     *bool
+}
+
+var _ proxylib.Parser = &Parser{}
+var t = true
+var f = false
+
+// OnData parses binary memcached data
+func (p *Parser) OnData(reply, endStream bool, dataBuffers [][]byte, offset int) (proxylib.OpType, int) {
+	if p.isBinary == nil {
+		remainingOffset := offset
```

Comment:
Issued https://github.com/cilium/cilium/pull/5730 which both simplifies and documents the Parser interface. Would be good for that to go in, as it would simplify this alot. For example, the only way there is no input will be if `len(dataBuffers) == 0` due to a change that makes sure none of the `[]byte` is empty.

Reply:
done